github/Requester.py

```diff
@@ -349,6 +365,47 @@ def __init__(
         self.__userAgent = user_agent
         self.__verify = verify
 
+    def _must_refresh_token(self) -> bool:
+        """Check if it is time to refresh the API token gotten from the GitHub app installation"""
+        if not self.__installation_authorization:
+            return False
+        return (
+            self.__installation_authorization.expires_at
+            < datetime.datetime.utcnow()
+            + datetime.timedelta(seconds=ACCESS_TOKEN_REFRESH_THRESHOLD_SECONDS)
+        )
+
+    def _get_installation_authorization(self):
+        assert (
+            self.__app_id is not None
+            and self.__app_private_key is not None
+            and self.__app_installation_id is not None
+        )
+        integration = GithubIntegration.GithubIntegration(
+            self.__app_id, self.__app_private_key
+        )
+        return integration.get_access_token(
+            self.__app_installation_id, permissions=self.__app_token_permissions
+        )
+
+    def _refresh_token_if_needed(self) -> None:
+        """Get a new access token from the GitHub app installation if the one we have is about to expire"""
+        if not self.__installation_authorization:
+            return
+        if self._must_refresh_token():
+            logging.debug("Refreshing access token")
+            self._refresh_token()
+
+    def _refresh_token(self) -> None:
+        """In the context of a GitHub app, refresh the access token"""
+        assert (
+            self.__app_id is not None
+            and self.__app_private_key is not None
+            and self.__app_installation_id is not None
+        )
+        self.__installation_authorization = self._get_installation_authorization()
+        self.__authorizationHeader = f"token {self.__installation_authorization.token}"
```

Comment:
This isn't thread-safe at the moment, and race conditions are quite likely to occur given the nature of the expiry. A few ideas:

1. an `RLock` around refreshing
2. put `_install_authorization` & `_authorization_header` in a thread-local, but that's more calls to the GH API even when there's no race.
3. do something like this using a tuple/namedtuple so if there is a refresh-race the instance attribute is assigned atomically. And change other uses of `__installation_authorization`/`__authorizationHeader` to reference the tuple.
```
install_authorization = self._get_installation_authorization()
authorization_header = f"token {self.__installation_authorization.token}"
# atomic assignment
self.__authorization = (install_authorization, authorization_header)
```